File: src/cmd/display.rs
use anyhow::Result;
use async_trait::async_trait;
use clap::Args;

use crate::db::repo::DisplayLevel;
use crate::debug;
use crate::repo::current::get_current_repo;

use super::{Command, ConfigArgs};

#[derive(Debug, Args)]
pub struct DisplayCommand {
    #[clap(flatten)]
    pub config: ConfigArgs,
}

#[async_trait]
impl Command for DisplayCommand {
    async fn run(self) -> Result<()> {
        debug!("[cmd] Run display command: {:?}", self);
        let ctx = self.config.build_ctx()?;
        ctx.lock()?;

        let repo = get_current_repo(ctx.clone())?;

        let remote = ctx.cfg.get_remote(&repo.remote)?;

        let text = match remote.icon {
            Some(ref icon) => format!("{icon} {}", repo.display_name(DisplayLevel::Owner)),
            None => repo.full_name(),
        };

        println!("{text}");
        Ok(())
    }

    fn complete_command() -> clap::Command {
        Self::augment_args(clap::Command::new("display"))
    }
}

File: cmd/display.go
package cmd

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/spf13/cobra"
)

func NewDisplay() *cobra.Command {
	var opts displayOptions
	c := &cobra.Command{
		Use:   "display",
		Short: "Display the current repository",

		Args: cobra.MaximumNArgs(1),

		ValidArgsFunction: DirCompletion,
	}

	return Build(c, &opts)
}

type displayOptions struct {
	path string
}

func (o *displayOptions) Complete(c *cobra.Command, args []string) error {
	if len(args) > 0 {
		path := args[0]
		path, err := filepath.Abs(path)
		if err != nil {
			return fmt.Errorf("failed to get the absolute path: %w", err)
		}

		path = strings.TrimSuffix(path, "/")
		o.path = path
	}
	return nil
}

func (o *displayOptions) Run(ctx *context.Context) error {
	if o.path != "" {
		ctx.WorkDir = o.path
	}
	repo, err := repoutils.MustGetCurrentRepo(ctx)
	if err != nil {
		return err
	}

	remoteConfig, err := ctx.GetRemote(repo.Remote)
	if err != nil {
		return err
	}

	format := ctx.Config.DisplayFormat

	output := strings.ReplaceAll(format, "{icon}", remoteConfig.Icon)
	output = strings.ReplaceAll(output, "{remote}", repo.Remote)
	output = strings.ReplaceAll(output, "{owner}", repo.Owner)
	output = strings.ReplaceAll(output, "{name}", repo.Name)

	fmt.Println(output)
	return nil
}
